xonsh/events.py

```diff
@@ -0,0 +1,113 @@
+"""
+Events for xonsh.
+
+In all likelihood, you want builtins.__xonsh_events__
+
+The best way to "declare" an event is something like::
+
+    __xonsh_events__.on_spam.__doc__ = "Comes with eggs"
+"""
+
+
+class Event(set):
+    """
+    A given event that handlers can register against.
+
+    Acts as a ``set`` for registered handlers.
+
+    Note that ordering is never guarenteed.
+    """
+    def __init__(self, doc=None):
+        self.__doc__ = doc
+
+    def handler(self, callable):
+        """
+        Registers a handler. It's suggested to use this as a decorator.
+        """
+        self.add(callable)
+        return callable
+
+    def calleach(self, *pargs, **kwargs):
+        """
+        The core handler caller that all others build on.
+
+        This works as a generator. Each handler is called in turn and its
+        results are yielded.
+
+        If the generator is interupted, no further handlers are called.
+
+        The caller may send() new positional arguments (eg, to implement
+        modifying semantics). Keyword arguments cannot be modified this way.
+        """
+        for handler in self:
```

Comment:
I can't imagine a use case where the handler couldn't fail fast instead of having an extra lambda (or worse, a full function).

cd, in particular, I don't agree with you. The two xontribs that would make use of the cd event so far (autoxsh, avox) have to call out to the file system to see if they can run or not. And the way avox is architected (all the filesystem calls are handled by `Vox`, so avox has no knowledge of how virtual environments are actually stored), it can't efficiently make use of a validator function.

In general, class-based commands could not make use of a validator function.